### PR TITLE
UI Functional Tests for launcher

### PR DESCRIPTION
**Pull request by sanketpathak**
What this PR do?
- Add UI tests for launcher

Issue is tracked here:
https://openshift.io/openshiftio/Openshift_io/plan/detail/520

### CODE REVIEW

File: src/test/page_objects/launcher/ui/create-applictionpage.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        await this.continueButton.clickWhenReady();
+    }
+
+    async cancelNewApplication(applicationName: string) {
+        await this.newApplication(applicationName);
+        await this.cancelSign.clickWhenReady();
+    }
+
+    private async newApplication(applicationName: string) {
+        console.log('-----oh------');
```
Comment (Raunak1203) [inline src/test/page_objects/launcher/ui/create-applictionpage.ts]: please remove console.log

File: src/test/page_objects/launcher/ui/create-applictionpage.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import * as ui from '../../../ui' ;
+import { element, by } from 'protractor';
+
+
+export class CreateApplication extends ui.BaseElement {
+    createApplication = new ui.BaseElement(this.$('.code-imports--step_title'), 'create application string');
+    importDialog = new ui.BaseElement(this.$('.imports-dialog'), 'create application dialog');
+    newApplicationName = new ui.TextInput(this.$('#projectName'), 'new application name');
+    // tslint:disable-next-line:max-line-length
```
Comment (Raunak1203) [inline src/test/page_objects/launcher/ui/create-applictionpage.ts]: please fix this

File: src/test/page_objects/launcher/ui/create-applictionpage.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    async cancelNewApplication(applicationName: string) {
+        await this.newApplication(applicationName);
+        await this.cancelSign.clickWhenReady();
+    }
+
+    private async newApplication(applicationName: string) {
+        console.log('-----oh------');
+        await this.createApplication.untilTextIsPresent('Create an Application');
+        await this.importDialog.ready();
+        console.log('+++gg+++');
```
Comment (Raunak1203) [inline src/test/page_objects/launcher/ui/create-applictionpage.ts]: please remove console.log

File: src/test/page_objects/launcher/ui/create-applictionpage.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        await this.newApplicationName.enterText(applicationName);
+        console.log('---gg---');
```
Comment (Raunak1203) [inline src/test/page_objects/launcher/ui/create-applictionpage.ts]: please remove console.log

File: src/test/spec/smoke_test.spec.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  });
+
+  beforeEach(async() => {
+    await launcher.ready();
+    let space = new support.SpaceName;
+    spaceName = space.newSpaceName();
+    appName = space.newApplicationName();
+  });
+
+  fit('should create new application', async() => {
```
Comment (Raunak1203) [inline src/test/spec/smoke_test.spec.ts]: remove f from fit

File: src/test/spec/smoke_test.spec.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    await launcher.appHeader.selectCreateSpace(' Create space ');
+    await launcher.createSpacePage.createNewSpace(spaceName);
+    await launcher.waitUntilUrlContains('/' + spaceName);
+    expect(await browser.getCurrentUrl()).toContain(spaceName);
+    await launcher.createApplicationPage.cancelNewApplication('testdemo');
+    await launcher.appHeader.untilTextIsPresent('Analyze');
+    expect (await browser.getTitle()).toContain('Analyze');
+  });
+
+  fit('should create new application', async() => {
```
Comment (Raunak1203) [inline src/test/spec/smoke_test.spec.ts]: remove f from fit

File: src/test/spec/smoke_test.spec.ts [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    await launcher.setupApp.newProjectBoosterOkIcon('Creating your project on OpenShift').untilDisplayed();
+    await launcher.setupApp.newProjectBoosterOkIcon('Setting up your build pipeline').untilDisplayed();
+    await launcher.setupApp.newProjectBoosterOkIcon('Configuring to trigger builds on Git pushes')
+      .untilDisplayed();
+    await launcher.setupApp.clickViewNewApplicationButton();
+    await launcher.appHeader.untilTextIsPresent('Analyze');
+    expect (await browser.getTitle()).toContain('Analyze');
+  });
+
+  xit('should import application', async() => {
```
Comment (Raunak1203) [inline src/test/spec/smoke_test.spec.ts]: remove xit

File: src/test/spec/smoke_test.spec.ts [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    await launcher.appHeader.selectCreateSpace(' Create space ');
+    await launcher.createSpacePage.createNewSpace(spaceName);
+    await launcher.waitUntilUrlContains('/' + spaceName);
+    expect(browser.getCurrentUrl()).toContain(spaceName);
+    await launcher.createApplicationPage.importAnApplication(appName);
+    await launcher.waitUntilUrlContains('/importapp/' + appName);
+    await launcher.importGitProvider.createNewApplication('test10');
+    await launcher.importSelectPipeline.createNewApplication();
+    await launcher.importApplication.createNewApplication();
```
Comment (Raunak1203) [inline src/test/spec/smoke_test.spec.ts]: please write assert statement

File: tests/protractor.conf.ts [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        },
+        summary: {
+          displayDuration: true
+        }
+      })
+    );
+    // Disable control flow
+    browser.ignoreSynchronization = true;
+    browser.baseUrl = browser.baseUrl + '/_home';
+    // browser.token = encodeURIComponent(JSON.stringify({
```
Comment (Raunak1203) [inline tests/protractor.conf.ts]: Do we need browser.token?

File: tests/run_e2e_tests.sh [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+# Exit handler
+trap cleanup EXIT
+
+# clone_fabric8_test() {
+#   log "Cloning fabric8-test to $temp_dir"
+#   git clone https://github.com/fabric8io/fabric8-test.git $temp_dir/fabric8-test
+# }
+
+# generate_db() {
```
Comment (Raunak1203) [inline tests/run_e2e_tests.sh]: Do we need this ?

File: src/test/page_objects/base.page.ts [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    await browser.get(this.url);
+
+    let urlNow = await browser.getCurrentUrl();
+    this.debug('now :', urlNow);
+  }
+
+  async login(username: string, password: string): Promise<void> {
+    this.log('base url set');
+    await browser.get(browser.baseUrl);
+    await browser.sleep(5000);
```
Comment (Raunak1203) [inline src/test/page_objects/base.page.ts]: please avoid browser.sleep()

File: src/test/page_objects/launcher/ui/app-launcher/select-pipeline.ts [hunk 12]
```diff
@@ -0,0 +1,16 @@
+import * as ui from '../../../../ui' ;
+
+
+export class SelectPipeline extends ui.BaseElement {
+    // tslint:disable-next-line:max-line-length
```
Comment (Raunak1203) [inline src/test/page_objects/launcher/ui/app-launcher/select-pipeline.ts]: fix this

**Comment by edewit** [comment]
seems these tests can't be run on the ngx-launcher project and be specific to test the integration with fabric8-ui. Wouldn't it be better to have them "live" there?

**Comment by Raunak1203** [comment]
> seems these tests can't be run on the ngx-launcher project and be specific to test the integration with fabric8-ui. Wouldn't it be better to have them "live" there?

@edewit we can store username and password in semaphoreci and run tests on each PR after unit tests?

**Comment by rgarg1** [comment]
@edewit We're trying to ensure that ngx-launcher PRs are of high quality, plus the changes do not break its integration with fabric8-ui. If these tests reside in fabric8-ui, PR validation won't be possible and any integration failure detected would be very late.

**Comment by edewit** [comment]
Okay, but how would that work? Is there a process that updates fabric8-ui with the unreleased PR and then runs these tests? On our side we use dependabot to create a PR when there is a release of ngx-launcher and run the integration tests at that time. But if you wanted to have this notification earlier we could also do it with a webhook. But I feel that these test could better be on fabric8-ui. Just because people that make changes on this repo don't necessarily have to run all the integration tests of all the projects depending on this.

**Comment by bartoszmajsak** [comment]
@rgarg1 

> PR validation won't be possible and any integration failure detected would be very late.

It certainly would be possible. We can have a webhook (if maintainers let us do so :)) and react on PR events to run our own tests in our own infra. This would let us know if incoming changes are affecting us, fabric8-ui, as the consumer of this upstream component. What @edewit is suggesting makes perfect sense for me.

**Comment by rgarg1** [comment]
@bartoszmajsak 

This approach sounds promising to me. Let's talk about it later today. Thanks @edewit

**Comment by bartoszmajsak** [comment]
> After a quick look on this PR it seems that you are duplicating a lot of work that was already done in e2e UI tests and you are making the same mistakes that we did.

Long live transparency and collaboration ;)

**Pull request closed without merging**